File: sender-mac/Sources/SenderMac/SignalingMessages.swift
import Foundation

enum SignalingRole: String, Codable {
    case sender
    case receiver
}

struct IceCandidatePayload: Codable, Sendable {
    let candidate: String
    let sdpMid: String?
    let sdpMLineIndex: Int?
}

enum SignalingServerMessage: Sendable {
    case joined(roomId: String, role: SignalingRole)
    case peerJoined(role: SignalingRole)
    case offer(sdp: String)
    case answer(sdp: String)
    case iceCandidate(IceCandidatePayload)
    case peerLeft(role: SignalingRole)
    case pong
    case error(message: String)
    case latencySync(sequence: Int64, senderTimeMs: Int64)
    case latencyEcho(sequence: Int64, senderTimeMs: Int64, receiverTimeMs: Int64)
    case unknown(type: String)
}

enum SignalingClientMessage: Sendable {
    case join(roomId: String, role: SignalingRole)
    case offer(sdp: String)
    case answer(sdp: String)
    case iceCandidate(IceCandidatePayload)
    case leave
    case ping
    case latencySync(sequence: Int64, senderTimeMs: Int64)
    case latencyEcho(sequence: Int64, senderTimeMs: Int64, receiverTimeMs: Int64)

    var logType: String {
        switch self {
        case .join:
            "join"
        case .offer:
            "offer"
        case .answer:
            "answer"
        case .iceCandidate:
            "ice-candidate"
        case .leave:
            "leave"
        case .ping:
            "ping"
        case .latencySync:
            "latency-sync"
        case .latencyEcho:
            "latency-echo"
        }
    }

    func jsonData() throws -> Data {
        try JSONSerialization.data(withJSONObject: jsonObject(), options: [])
    }

    private func jsonObject() -> [String: Any] {
        switch self {
        case .join(let roomId, let role):
            return ["type": "join", "roomId": roomId, "role": role.rawValue]
        case .offer(let sdp):
            return ["type": "offer", "sdp": sdp]
        case .answer(let sdp):
            return ["type": "answer", "sdp": sdp]
        case .iceCandidate(let candidate):
            var candidateObject: [String: Any] = [
                "candidate": candidate.candidate
            ]
            if let sdpMid = candidate.sdpMid {
                candidateObject["sdpMid"] = sdpMid
            }
            if let sdpMLineIndex = candidate.sdpMLineIndex {
                candidateObject["sdpMLineIndex"] = sdpMLineIndex
            }
            return [
                "type": "ice-candidate",
                "candidate": candidateObject
            ]
        case .leave:
            return ["type": "leave"]
        case .ping:
            return ["type": "ping"]
        case .latencySync(let sequence, let senderTimeMs):
            return ["type": "latency-sync", "sequence": sequence, "senderTimeMs": senderTimeMs]
        case .latencyEcho(let sequence, let senderTimeMs, let receiverTimeMs):
            return [
                "type": "latency-echo",
                "sequence": sequence,
                "senderTimeMs": senderTimeMs,
                "receiverTimeMs": receiverTimeMs
            ]
        }
    }
}

enum SignalingMessageDecoder {
    static func decode(text: String) -> SignalingServerMessage {
        guard let data = text.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = object["type"] as? String else {
            return .unknown(type: "invalid-json")
        }

        switch type {
        case "joined":
            return .joined(
                roomId: object["roomId"] as? String ?? "",
                role: SignalingRole(rawValue: object["role"] as? String ?? "") ?? .sender
            )
        case "peer-joined":
            return .peerJoined(role: SignalingRole(rawValue: object["role"] as? String ?? "") ?? .receiver)
        case "offer":
            return .offer(sdp: object["sdp"] as? String ?? "")
        case "answer":
            return .answer(sdp: object["sdp"] as? String ?? "")
        case "ice-candidate":
            return .iceCandidate(parseCandidate(object["candidate"]))
        case "peer-left":
            return .peerLeft(role: SignalingRole(rawValue: object["role"] as? String ?? "") ?? .receiver)
        case "pong":
            return .pong
        case "error":
            return .error(message: object["message"] as? String ?? "unknown error")
        case "latency-sync":
            return .latencySync(
                sequence: int64(object["sequence"]),
                senderTimeMs: int64(object["senderTimeMs"])
            )
        case "latency-echo":
            return .latencyEcho(
                sequence: int64(object["sequence"]),
                senderTimeMs: int64(object["senderTimeMs"]),
                receiverTimeMs: int64(object["receiverTimeMs"])
            )
        default:
            return .unknown(type: type)
        }
    }

    private static func parseCandidate(_ value: Any?) -> IceCandidatePayload {
        guard let object = value as? [String: Any] else {
            return IceCandidatePayload(candidate: "", sdpMid: nil, sdpMLineIndex: nil)
        }

        return IceCandidatePayload(
            candidate: object["candidate"] as? String ?? "",
            sdpMid: object["sdpMid"] as? String,
            sdpMLineIndex: object["sdpMLineIndex"] as? Int
        )
    }

    private static func int64(_ value: Any?) -> Int64 {
        if let int = value as? Int {
            return Int64(int)
        }
        if let int64 = value as? Int64 {
            return int64
        }
        if let double = value as? Double {
            return Int64(double)
        }
        return 0
    }
}

File: receiver-quest/Assets/Scripts/Signaling/SignalingMessages.cs
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace X5Quest.Signaling
{
    public enum SignalingRole
    {
        Sender,
        Receiver
    }

    [Serializable]
    public sealed class IceCandidatePayload
    {
        public string candidate;
        public string sdpMid;
        public int? sdpMLineIndex;
    }

    public enum SignalingMessageType
    {
        Joined,
        PeerJoined,
        Offer,
        Answer,
        IceCandidate,
        PeerLeft,
        Pong,
        Error,
        LatencySync,
        LatencyEcho,
        Unknown
    }

    public sealed class SignalingMessage
    {
        public SignalingMessageType Type { get; private set; }
        public string RawType { get; private set; }
        public string RoomId { get; private set; }
        public SignalingRole Role { get; private set; }
        public string Sdp { get; private set; }
        public IceCandidatePayload Candidate { get; private set; }
        public string ErrorMessage { get; private set; }
        public long Sequence { get; private set; }
        public long SenderTimeMs { get; private set; }
        public long ReceiverTimeMs { get; private set; }

        public static SignalingMessage Parse(string json)
        {
            var obj = JObject.Parse(json);
            var type = obj.Value<string>("type") ?? "unknown";
            var message = new SignalingMessage { RawType = type, Type = MapType(type) };

            message.RoomId = obj.Value<string>("roomId") ?? string.Empty;
            message.Role = ParseRole(obj.Value<string>("role"));
            message.Sdp = obj.Value<string>("sdp") ?? string.Empty;
            message.ErrorMessage = obj.Value<string>("message") ?? string.Empty;
            message.Sequence = obj.Value<long?>("sequence") ?? 0;
            message.SenderTimeMs = obj.Value<long?>("senderTimeMs") ?? 0;
            message.ReceiverTimeMs = obj.Value<long?>("receiverTimeMs") ?? 0;

            var candidate = obj["candidate"];
            if (candidate != null)
            {
                message.Candidate = candidate.ToObject<IceCandidatePayload>();
            }

            return message;
        }

        public static string Join(string roomId)
        {
            return JsonConvert.SerializeObject(new
            {
                type = "join",
                roomId,
                role = "receiver"
            });
        }

        public static string Answer(string sdp)
        {
            return JsonConvert.SerializeObject(new { type = "answer", sdp });
        }

        public static string IceCandidate(IceCandidatePayload candidate)
        {
            return JsonConvert.SerializeObject(new { type = "ice-candidate", candidate });
        }

        public static string Leave()
        {
            return JsonConvert.SerializeObject(new { type = "leave" });
        }

        public static string Ping()
        {
            return JsonConvert.SerializeObject(new { type = "ping" });
        }

        public static string LatencyEcho(long sequence, long senderTimeMs, long receiverTimeMs)
        {
            return JsonConvert.SerializeObject(new
            {
                type = "latency-echo",
                sequence,
                senderTimeMs,
                receiverTimeMs
            });
        }

        private static SignalingMessageType MapType(string type)
        {
            switch (type)
            {
                case "joined": return SignalingMessageType.Joined;
                case "peer-joined": return SignalingMessageType.PeerJoined;
                case "offer": return SignalingMessageType.Offer;
                case "answer": return SignalingMessageType.Answer;
                case "ice-candidate": return SignalingMessageType.IceCandidate;
                case "peer-left": return SignalingMessageType.PeerLeft;
                case "pong": return SignalingMessageType.Pong;
                case "error": return SignalingMessageType.Error;
                case "latency-sync": return SignalingMessageType.LatencySync;
                case "latency-echo": return SignalingMessageType.LatencyEcho;
                default: return SignalingMessageType.Unknown;
            }
        }

        private static SignalingRole ParseRole(string role)
        {
            return role == "sender" ? SignalingRole.Sender : SignalingRole.Receiver;
        }
    }
}

